### ISSUE TITLE
proposal: `no_wildcard_variable_uses`

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#4392]
## Description

Do not use wildcard parameters or variables.

## Details

There's a proposal [to change the language](https://github.com/dart-lang/language/blob/master/working/wildcards/feature-specification.md) so that (at least) parameters and local variables named `_` (e.g., wildcards) become non-binding. When that happens, existing code that uses wildcard parameters or variables will break. 

Enabling this lint now will make it less breaking to make `_` non-binding in a future Dart release.

## Kind

Error

## Bad Examples

```dart
var _ = 1;
print(_); // LINT
```

```dart
void f(int __) {
  print(__); // LINT multiple underscores too
}
```

## Good Examples 

```dart
for (var _ in [1, 2, 3]) count++;
```

```dart
try {
  ...
} catch (_) {
  ...
}
```

```dart
var [a, _, b, _] = [1, 2, 3, 4];
```

## Discussion

https://github.com/dart-lang/sdk/issues/51221

cc @dart-lang/language-team

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by munificent** [comment]
I know I suggested this name, but now that I think about it, it would probably be better for the name to emphasize that it's *using* a variable whose name is a wildcard that's the issue, not *declaring* one. Maybe "do_not_use_wildcard"?

**Comment by pq** [comment]
Ah, right. It's a good point. I updated to `no_wildcard_variable_uses` (since we have more `no_`s than `do_not_`s) but `do_not_use_wildcard_variables` would be fine too.

Open to other ideas too!

/cc @bwilkerson

**Comment by pq** [comment]
Given the support in https://github.com/dart-lang/sdk/issues/51221, I'm going to transition this to accepted and offer a simple implementation.  (We don't have to land that obviously.)
